rust-runtime/aws-smithy-async/src/test_util/tick_advance_sleep.rs

```diff
@@ -146,22 +151,25 @@ impl TickAdvanceTime {
     pub async fn tick(&self, duration: Duration) {
         let time = self.inner.get().now + duration;
 
-        let mut presenting = self.inner.get_mut().take_presenting_before(time);
-        while !presenting.is_empty() {
-            // Tick to each individual sleep time and yield the runtime so that any
-            // futures waiting on a sleep run before futures waiting on a later sleep.
-            for mut sleep in presenting {
-                // Make sure the time is always accurate for async code that runs
-                // after completing the sleep.
-                self.inner.get_mut().now = sleep.presents_at;
-
-                let _ = sleep.notify.take().unwrap().send(());
-                tokio::task::yield_now().await;
-            }
+        // Tick to each individual sleep time and yield the runtime so that any
+        // futures waiting on a sleep run before futures waiting on a later sleep.
+        //
+        // We also need to recheck the list of queued sleeps every iteration since
+        // unblocked tasks could have queued up more sleeps, and these sleeps may also
+        // need to present before ones that were previously queued.
+        loop {
+            // Can't do `while let` since that holds the lock open
+            let Some(mut presenting) = self.inner.get_mut().next_presenting(time) else {
+                break;
+            };
+
+            // Make sure the time is always accurate for async code that runs
+            // after completing the sleep.
+            self.inner.get_mut().now = presenting.presents_at;
 
-            // Unblocked tasks could have queued some more sleeps within the duration,
-            // so we need to check those as well.
-            presenting = self.inner.get_mut().take_presenting_before(time);
+            // Notify the sleep, and then yield to let work blocked on that sleep to proceed
+            let _ = presenting.notify.take().unwrap().send(());
```

Comment:
so the only issue with this abstraction is that you don't have a guarantee that the code has _actually_ advanced past the sleep point (other than a "cross fingers") in `yield_now`. I suspect this MAY cause very occasionally flaky tests (or non flaky tests that rely on tokio implementation details)

Reply:
Yeah, it does break down for very large numbers of sleeps. For example, I tried having 100 tasks blocked on 100 sleeps, and not all the tasks were visited after unblocking the 100 sleeps before the TickAdvanceTime proceeded with the next sleep.

In practice though, I haven't observed any issues with real world tests yet. It would be cool for tests to be able to wait until Tokio's task queue is completely blocked though.